File: framework/shm_layout.zig
//! Shared memory layout primitives — the wire format common to every
//! SHM transport (shm_bus for sidecar CALL/RESULT, worker_dispatch for
//! background workers).
//!
//! **One source of truth, referenced by name.** Both transports declare
//! their own SlotHeader (they diverge on the trailing byte: shm_bus
//! carries an explicit slot_state enum, worker_dispatch does not), but
//! the shared leading 24 bytes, the RegionHeader, and the CRC convention
//! live here. A header-offset bug once had to be fixed twice — see
//! commit 7fe537e — which is what this module exists to prevent.
//!
//! Pair-assert via `assert_slot_header_layout(T)`: each transport's
//! SlotHeader calls it in a comptime block, which proves the common
//! fields are at the canonical offsets. Offsets match the cross-language
//! contract in packages/vectors/shm_layout.json.
//!
//! TB reference: vsr/checksum.zig — one named primitive, reused by name
//! from every call site. No generic container, no wrapper around the
//! extern struct. Just the substrate.

const std = @import("std");
const assert = std.debug.assert;
const Crc32 = std.hash.crc.Crc32;

/// RegionHeader — 64-byte capacity metadata at the start of every SHM
/// region. Identical layout for every transport. Readers (including the
/// TS sidecar) interpret this header first to learn slot_count and
/// frame_max before mapping the full region.
///
/// Field default policy (deliberately mixed):
///   - `slot_count` and `frame_max` have **no defaults** — they carry
///     semantic meaning per region and must be set explicitly at the
///     call site (TB: "explicitly pass options... instead of relying
///     on the defaults").
///   - `_reserved` and `_pad` **default to zero** — the protocol
///     requires them zero, and mmap'd regions come pre-zeroed by
///     @memset. Defaults match the wire contract.
///
/// This isn't an inconsistency; it's the right behavior per field. The
/// invariant "semantic fields require caller intent, reserved bytes are
/// always zero" is stronger than a blanket all-or-nothing default rule.
pub const RegionHeader = extern struct {
    slot_count: u16,
    _reserved: u16 = 0,
    frame_max: u32,
    _pad: [56]u8 = [_]u8{0} ** 56,

    comptime {
        assert(@sizeOf(@This()) == 64);
        assert(@offsetOf(@This(), "slot_count") == 0);
        assert(@offsetOf(@This(), "_reserved") == 2);
        assert(@offsetOf(@This(), "frame_max") == 4);
        assert(@offsetOf(@This(), "_pad") == 8);
    }
};

/// Assert that a transport-specific SlotHeader places the shared leading
/// fields at the canonical offsets. Every SlotHeader calls this in a
/// comptime block — if the layout drifts, the build fails.
///
/// Offsets are frozen by the cross-language contract in
/// packages/vectors/shm_layout.json. The TS sidecar reads these offsets
/// directly; drifting them silently breaks dispatch.
pub fn assert_slot_header_layout(comptime T: type) void {
    comptime {
        assert(@sizeOf(T) == 64);
        assert(@offsetOf(T, "server_seq") == 0);
        assert(@offsetOf(T, "sidecar_seq") == 4);
        assert(@offsetOf(T, "request_len") == 8);
        assert(@offsetOf(T, "response_len") == 12);
        assert(@offsetOf(T, "request_crc") == 16);
        assert(@offsetOf(T, "response_crc") == 20);
    }
}

/// Compute the CRC32 of a frame: CRC(u32 LE payload_len ++ payload).
///
/// The length-prefix then payload convention is part of the
/// cross-language contract (packages/vectors/shm_layout.json:
/// crc_convention). A corrupted length produces a CRC mismatch rather
/// than a garbage read — matches TB's convention for disk writes.
///
/// `inline` because this is called on every CALL write, every RESULT
/// read, and every worker dispatch — the function body is ~5 instructions
/// plus the Crc32 loop and inlining lets the compiler fold the `len`
/// bytes into the CRC state directly.
pub inline fn crc_frame(len: u32, payload: []const u8) u32 {
    // The caller's buffer must hold at least `len` bytes — otherwise
    // we'd CRC garbage beyond the declared frame.
    assert(payload.len >= len);

    var crc = Crc32.init();
    crc.update(std.mem.asBytes(&len));
    crc.update(payload[0..len]);
    return crc.final();
}

// =====================================================================
// Tests
// =====================================================================

test "RegionHeader size and offsets" {
    try std.testing.expectEqual(@as(usize, 64), @sizeOf(RegionHeader));
    try std.testing.expectEqual(@as(usize, 0), @offsetOf(RegionHeader, "slot_count"));
    try std.testing.expectEqual(@as(usize, 4), @offsetOf(RegionHeader, "frame_max"));
}

test "crc_frame is a deterministic LE-len ++ payload CRC" {
    // Empty payload: CRC over four zero bytes.
    try std.testing.expectEqual(Crc32.hash(&[_]u8{ 0, 0, 0, 0 }), crc_frame(0, &[_]u8{}));

    // One-byte payload: CRC over 4-byte length ++ payload byte.
    const one = [_]u8{0xAB};
    try std.testing.expectEqual(
        Crc32.hash(&[_]u8{ 1, 0, 0, 0, 0xAB }),
        crc_frame(1, &one),
    );

    // Length prefix matters: same payload, different declared length
    // produces a different CRC (guards against length-forgery).
    const two = [_]u8{ 0xAB, 0xCD };
    const c1 = crc_frame(1, &two);
    const c2 = crc_frame(2, &two);
    try std.testing.expect(c1 != c2);
}

test "assert_slot_header_layout accepts a canonical SlotHeader" {
    // Transport-shaped: 6 u32s + 1 u8 state + 39 bytes pad = 64 bytes.
    const Canonical = extern struct {
        server_seq: u32,
        sidecar_seq: u32,
        request_len: u32,
        response_len: u32,
        request_crc: u32,
        response_crc: u32,
        slot_state: u8,
        _pad: [39]u8,
    };
    comptime assert_slot_header_layout(Canonical);
}

test "assert_slot_header_layout accepts a pad-only SlotHeader" {
    // worker_dispatch shape: 6 u32s + 40 bytes pad = 64 bytes.
    const PadOnly = extern struct {
        server_seq: u32,
        sidecar_seq: u32,
        request_len: u32,
        response_len: u32,
        request_crc: u32,
        response_crc: u32,
        _pad: [40]u8,
    };
    comptime assert_slot_header_layout(PadOnly);
}

File: packages/ts/src/shm_layout.ts
// Shared memory layout primitives — the TS mirror of framework/shm_layout.zig.
//
// One source of truth, referenced by name. Every CALL/RESULT write and
// every CRC check routes through `crcFrame` and `HEADER_OFFSETS` here
// so the Zig↔C↔TS wire contract can only drift in one place.
//
// Pair assertion: `selfTest` runs on module load and recomputes the
// canonical cross-language test vector. If it diverges from the Zig
// or C implementation the process aborts — matches the CRC self-test
// in shm.c (init-time) and worker_dispatch.zig ("CRC32 cross-language
// test vector" test).

import { crc32 } from "node:zlib";

/// Length-prefix CRC convention: CRC32(LE u32 len ++ payload[0..len]).
///
/// A corrupted length produces a CRC mismatch rather than a garbage
/// read. Matches Zig's `framework/shm_layout.zig:crc_frame` and C's
/// `packages/ts/native/shm.c:compute_crc`.
export function crcFrame(len: number, payload: Uint8Array): number {
  if ((len >>> 0) !== len) {
    throw new Error(`crcFrame: len must be a u32, got ${len}`);
  }
  if (payload.length < len) {
    throw new Error(`crcFrame: payload.length=${payload.length} < len=${len}`);
  }
  const lenBuf = Buffer.alloc(4);
  lenBuf.writeUInt32LE(len, 0);
  return crc32(Buffer.concat([lenBuf, payload.subarray(0, len)])) >>> 0;
}

/// Canonical field offsets inside a slot header. Mirrors the contract
/// in `packages/vectors/shm_layout.json` and the pair-assertion in
/// `framework/shm_layout.zig:assert_slot_header_layout`.
export const HEADER_OFFSETS = Object.freeze({
  SERVER_SEQ: 0,
  SIDECAR_SEQ: 4,
  REQUEST_LEN: 8,
  RESPONSE_LEN: 12,
  REQUEST_CRC: 16,
  RESPONSE_CRC: 20,
  SLOT_STATE: 24, // shm_bus only; worker_dispatch reserves this byte.
  SLOT_HEADER_SIZE: 64,
});

/// Region header offsets. 64 bytes of capacity metadata at offset 0
/// of every SHM region.
export const REGION_OFFSETS = Object.freeze({
  SLOT_COUNT: 0,      // u16 LE
  FRAME_MAX: 4,       // u32 LE
  REGION_HEADER_SIZE: 64,
});

/// Canonical cross-language CRC test vector: payload "hello" (5 bytes),
/// framed as [0x05, 0x00, 0x00, 0x00, 0x68, 0x65, 0x6c, 0x6c, 0x6f].
/// Expected CRC32 = 0x5CAC007A.
///
/// Identical vector is asserted in:
///   - framework/worker_dispatch.zig : test "CRC32 cross-language test vector"
///   - packages/ts/native/shm.c      : verify_crc_table (init-time)
///
/// If any language's implementation drifts, this three-way check fails.
export const CROSS_LANGUAGE_VECTOR = Object.freeze({
  payload: "hello",
  expectedCrc: 0x5CAC007A,
});

/// Startup pair-assertion. Run once at module load; throws (not silent
/// return) if the CRC doesn't match, since a silent mismatch would
/// drop every SHM frame with no visible error.
function selfTest(): void {
  const payload = new TextEncoder().encode(CROSS_LANGUAGE_VECTOR.payload);
  const actual = crcFrame(payload.length, payload);
  if (actual !== CROSS_LANGUAGE_VECTOR.expectedCrc) {
    throw new Error(
      `shm_layout: CRC self-test failed. Got 0x${actual.toString(16).toUpperCase()}, ` +
      `expected 0x${CROSS_LANGUAGE_VECTOR.expectedCrc.toString(16).toUpperCase()}. ` +
      `node:zlib crc32 has diverged from the Zig/C contract — all SHM frames would fail.`,
    );
  }
}

selfTest();
